compose/ui/ui/src/webTest/kotlin/CanvasBasedWindowTests.kt

```diff
@@ -77,19 +89,93 @@ class CanvasBasedWindowTests {
 
-        canvasElement.dispatchEvent(createTypedEvent())
+        canvasElement.dispatchEvent(createEventShouldNotBePrevented())
         assertEquals(2, stack.size)
-        assertTrue( stack.last())
-        assertEquals("c", changedValue)
+        assertFalse(stack.last())
 
-        canvasElement.dispatchEvent(createEventShouldNotBePrevented())
+        // copy shortcut should not be prevented (we let browser create a corresponding event)
+        canvasElement.dispatchEvent(createCopyKeyboardEvent())
         assertEquals(3, stack.size)
         assertFalse(stack.last())
     }
+
+    @Test
+    // https://github.com/JetBrains/compose-multiplatform/issues/3644
+    fun keyMappingIsValid() {
+        if (isHeadlessBrowser()) return
+
+        val canvasElement = document.createElement("canvas") as HTMLCanvasElement
+        canvasElement.setAttribute("id", canvasId)
+        document.body!!.appendChild(canvasElement)
+
+        val fr = FocusRequester()
+        var mapping = ""
+        CanvasBasedWindow(canvasElementId = canvasId) {
+            Box(Modifier.size(1000.dp).background(Color.Red).focusRequester(fr).focusTarget().onKeyEvent {
+                mapping = it.key.toString()
+                println(mapping)
+                false
+            }) {
+                Text("Try to press different keys and look at the console...")
+            }
+            SideEffect {
+                fr.requestFocus()
+            }
+        }
+
+        canvasElement.dispatchEvent(createTypedEvent('t'))
+        assertEquals("Key keyCode: 84", mapping)
+
+        canvasElement.dispatchEvent(createTypedEvent('6'))
+        assertEquals("Key keyCode: 54", mapping)
+    }
+
+    @Test
+    // https://github.com/JetBrains/compose-multiplatform/issues/2296
+    fun onPreviewKeyEventShouldWork() {
+        if (isHeadlessBrowser()) return
+        val canvasElement = document.createElement("canvas") as HTMLCanvasElement
+        canvasElement.setAttribute("id", canvasId)
+        document.body!!.appendChild(canvasElement)
+
+        val fr = FocusRequester()
+        val textValue = mutableStateOf("")
+        var lastKeyEvent: KeyEvent? = null
+        var stopPropagation = true
+
+        CanvasBasedWindow(canvasElementId = canvasId) {
+            TextField(
+                value = textValue.value,
+                onValueChange = { textValue.value = it },
+                modifier = Modifier.fillMaxSize().focusRequester(fr).onPreviewKeyEvent {
+                    lastKeyEvent = it
+                    return@onPreviewKeyEvent stopPropagation
+                }
+            )
+            SideEffect {
+                fr.requestFocus()
+            }
+        }
+
+        canvasElement.dispatchEvent(createTypedEvent('t'))
+        canvasElement.dispatchEvent(createTypedEvent('e'))
+        canvasElement.dispatchEvent(createTypedEvent('s'))
+        canvasElement.dispatchEvent(createTypedEvent('t'))
+        assertEquals(Key.T, lastKeyEvent!!.key)
```

Comment:
I might be missing something but how we are sure that this is the last "t"-event, not the first one?

Reply:
Changed the tests to avoid this potential issue.